File: Codes/Problem2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 *   Space Complexity - Without Recursive Stack - O(n)
 *                      With Recursive Stack - O(log(n))
 */

class Solution {
    int sum=0;
    public int sumNumbers(TreeNode root) {
        helper(root,0);
        return sum;
    }

    private void helper(TreeNode curr, int currSum){

        if(curr == null)
            return;

        if(curr.left == null && curr.right == null){
            sum += ((currSum*10) + curr.val);
            return;
        }
        int val = currSum*10 + curr.val;
        helper(curr.left, val);
        helper(curr.right,val);
    }
}

File: problem2.cpp
/**
// Time Complexity : O(N)
// Space Complexity : O(H)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : condition when one of the child is null and other is not.


// Your code here along with comments explaining your approach
*/



#include<iostream>
#include<vector>

using namespace std;


struct TreeNode{
    int val;
    TreeNode* left;
    TreeNode* right;
    TreeNode(int x,TreeNode* left,TreeNode* right):val(x),left(left),right(right){};
    TreeNode():val(0),left(nullptr),right(nullptr){};
    TreeNode(int x):val(x),left(nullptr),right(nullptr){};

};


class Solution {
public:
    int sum{};
    void helper(TreeNode* root,int curr_num){
        //base
        if(root == nullptr) return; // there is a possibility either left or right child is null;
        curr_num = curr_num*10+root->val;
        if(root->left == nullptr && root->right == nullptr){
            sum+= curr_num;
            return;
        }

        helper(root->left,curr_num);
        helper(root->right,curr_num);

    }
    int sumNumbers(TreeNode* root) {
        helper(root,0);
        return sum;
    }
};